src/server/mcp.ts

```diff
@@ -46,6 +47,64 @@ import { UriTemplate, Variables } from "../shared/uriTemplate.js";
 import { RequestHandlerExtra } from "../shared/protocol.js";
 import { Transport } from "../shared/transport.js";
 
+type ParametersJsonSchema = { type: 'object' } & Partial<JsonSchema7ObjectType>;
+
+/**
+ * Helper to check if an object is a Zod schema (ZodRawShape)
+ */
+const isZodRawShape = (obj: unknown): obj is ZodRawShape => {
+  if (typeof obj !== "object" || obj === null) return false;
+  // Check that at least one property is a ZodType instance
+  return Object.values(obj as object).some((v) => v instanceof ZodType);
+};
+
+const isJsonSchemaObject = (obj: unknown): obj is ParametersJsonSchema => {
+  return Boolean(
+    obj && typeof obj === "object" && "type" in obj && typeof obj.type === "string"
```

Comment:
Since `type` is not a required keyword for a JSON schema and it would be more robust not to ever infer a JSON schema version, can we just look for [`$schema`](https://json-schema.org/draft/2020-12/draft-bhutton-json-schema-01#section-8.1.1) instead?